### ISSUE TITLE
Not showing updated value after settings saved for options page field

### ISSUE THREAD

**Issue by byananta** [CMB2/CMB2#1102]
I have added a option page with a field. When i am trying to save the data the has been stored properly but it is not showing in the text field. Please help me.

### Steps to reproduce (I have confirmed I can reproduce this issue on the trunk branch):

1.
2.

### CMB2 Field Registration Code:

```php
add_action( 'cmb2_admin_init', 'yourprefix_register_demo_metabox' );
function yourprefix_register_demo_metabox() {

	$cmb = new_cmb2_box( array(
		// Box Config...
	) );

	$cmb->add_field( array(
		// Field Config...
	) );

	// Additional fields...
}
```

**Comment by byananta** [comment]
add_action('cmb2_admin_init', 'testing'  ); 
 public function testing() {
	/**
	 * Registers options page menu item and form.
	 */
	$cmb = new_cmb2_box( array(
		'id'           => 'ap_option_metabox',
		'title'        => esc_html__( 'Site Options', 'myprefix' ),
		'object_types' => array( 'options-page' ),
		/*
		 * The following parameters are specific to the options-page box
		 * Several of these parameters are passed along to add_menu_page()/add_submenu_page().
		 */
		'option_key'      => 'ap_options',
		'menu_title'      => esc_html__( 'Options', 'myprefix' ),
		'parent_slug'     => 'edit.php',
		//'capability'      => 'manage_options',
	) );


 $cmb->add_field( array(
 	'name' => 'Image Caption',
 	'id'   => 'image_caption',
 	'type' => 'text',
 ) );
 //var_dump(get_option('ap_options'));
}

**Comment by tw2113** [comment]
Presently working for me, using provided code as is, except for a renamed function/callback.

Is the code in your 2nd reply above what you're using exactly? or is it just a quick example? Asking because of the `public` keyword outside the context of a class.

**Comment by byananta** [comment]
Hi, thanks for your reply. Above code is just a reference.. I have found the problem when the option page under a submenu like edit.php or under any custom post type at that time the value is not poputating after save the data. But its working in xampp but not in my remote hosting. Xampp php version 7.1.4 and hosting php version is 5.3.33. But i am getting the option but not in the field value..

**Comment by byananta** [comment]
When i am using the parent_slug then the problem if i remove that then its working.

**Comment by tw2113** [comment]
Can you please provide the exact code you're using for your case, so that we can better make sure we're debugging/reviewing under matching conditions?

The PHP version I was using earlier with my tests and the reference code above was PHP 5.3.x as well, and no issue.

Something else must be up.

**Comment by byananta** [comment]
This code is working for me. When i am trying to make it under another menu or post type then the value is not populating after saving the data. This is the code which i am using.

add_action('cmb2_admin_init', array($this, 'testing' ) ); 
 public function testing(){
    $cmb_options = new_cmb2_box( array(
  		'id'           => 'ap_option_metabox',
  		'title'        => esc_html__( 'Site Options', 'myprefix' ),
  		'object_types' => array( 'options-page' ),
  		/*
  		 * The following parameters are specific to the options-page box
  		 * Several of these parameters are passed along to add_menu_page()/add_submenu_page().
  		 */
  		'option_key'      => 'ap_options', // The option key and admin menu page slug.
  		// 'icon_url'        => 'dashicons-palmtree', // Menu icon. Only applicable if 'parent_slug' is left empty.
  		// 'menu_title'      => esc_html__( 'Options', 'myprefix' ), // Falls back to 'title' (above).
  		// 'parent_slug'     => 'themes.php', // Make options page a submenu item of the themes menu.
  		// 'capability'      => 'manage_options', // Cap required to view options-page.
  		// 'position'        => 1, // Menu position. Only applicable if 'parent_slug' is left empty.
  		// 'admin_menu_hook' => 'network_admin_menu', // 'network_admin_menu' to add network-level options page.
  		// 'display_cb'      => false, // Override the options-page form output (CMB2_Hookup::options_page_output()).
  		// 'save_button'     => esc_html__( 'Save Theme Options', 'myprefix' ), // The text for the options-page save button. Defaults to 'Save'.
  	) );
  	/*
  	 * Options fields ids only need
  	 * to be unique within this box.
  	 * Prefix is not needed.
  	 */
  	$cmb_options->add_field( array(
  		'name' => __( 'Test Text', 'myprefix' ),
  		'desc' => __( 'field description (optional)', 'myprefix' ),
  		'id'   => 'test_text',
  		'type' => 'text',
  		'default' => 'Default Text',
  	) );
  	$cmb_options->add_field( array(
  		'name'    => __( 'Test Color Picker', 'myprefix' ),
  		'desc'    => __( 'field description (optional)', 'myprefix' ),
  		'id'      => 'test_colorpicker',
  		'type'    => 'colorpicker',
  		'default' => '#bada55',
  	) );
  }

**Comment by byananta** [comment]
This is not working.
![image](https://user-images.githubusercontent.com/33488818/37073493-4f9e8b10-21ed-11e8-934a-df161ef5b990.png)

**Comment by tw2113** [comment]
https://cloudup.com/c74_1JuMuYY

Still not presently having any issue getting things to save and show afterwards. This is the code I'm using:

https://gist.github.com/tw2113/68daffdb1dc2e8652d62271f16e98300

Tried as both a mu-plugin as well as in twentyseventeen's functions.php file.

To clarify, I'm not saying you're not having a problem with this as a whole, but thus far, it's feeling like something related to your current website, rather than CMB2 itself.

**Comment by byananta** [comment]
This is happening with my site because i have tested this with another hosting. It is working there.

**Comment by tw2113** [comment]
Best I can offer at that point then is deactivating plugins one by one until it starts working like expected. At that point, you likely have your conflicting culprit and can further debug from there with a stronger lead.

**Comment by byananta** [comment]
Actually i am using cmb2 in my plugin. When i am installing this plugin in another wordpress then it is working fine.. i have also removed all the other plugins but same problem. I am trying to findout which code is creating the problem..

**Comment by tw2113** [comment]
Keep us updated, I'm getting kind of curious what the conflict is.

**Comment by tw2113** [comment]
@byananta did you ever get this figured out? At least if you recall?

**Comment by byananta** [comment]
I did not find exact cause. But same code was working in another wordpress
setup.

Thanks,
Ananta

**Comment by tw2113** [comment]
I'm going to call it an anomaly then and close the issue for the time being. If it crops up again by yourself or someone else, we can re-investigate.

**Comment by TheNewGuy16** [comment]
Hey I am still having this issue.  If I put the option page submenu under a custom post type menu the fields do not get save and shown in the admin area. But if I don't add a parent slug the fields save.  I deactivated other plugins but still no luck.

`// ############ Add Options Page ################
add_action( 'cmb2_admin_init', 'directors_register_options_submenu' );

function directors_register_options_submenu() {
	/**
	 * Registers options page menu item and form.
	 */
	$cmb = new_cmb2_box( array(
		'id'           => 'directors_options_submenu_page',
		'title'        => esc_html__( 'Options', 'cmb2' ),
		'object_types' => array( 'options-page' ),
		/*
		 * The following parameters are specific to the options-page box
		 * Several of these parameters are passed along to add_menu_page()/add_submenu_page().
		 */
		'option_key'      => 'directors_page_options', // The option key and admin menu page slug.
		// 'icon_url'        => '', // Menu icon. Only applicable if 'parent_slug' is left empty.
		// 'menu_title'      => esc_html__( 'Options', 'cmb2' ), // Falls back to 'title' (above).
		'parent_slug'     => 'edit.php?post_type=director_item', // Make options page a submenu item of the themes menu.
		// 'capability'      => 'manage_options', // Cap required to view options-page.
		// 'position'        => 1, // Menu position. Only applicable if 'parent_slug' is left empty.
		// 'admin_menu_hook' => 'network_admin_menu', // 'network_admin_menu' to add network-level options page.
		// 'display_cb'      => false, // Override the options-page form output (CMB2_Hookup::options_page_output()).
		// 'save_button'     => esc_html__( 'Save Theme Options', 'cmb2' ), // The text for the options-page save button. Defaults to 'Save'.
		// 'disable_settings_errors' => true, // On settings pages (not options-general.php sub-pages), allows disabling.
		// 'message_cb'      => 'yourprefix_options_page_message_callback',
	) );
	$cmb->add_field( array(
		'name'    => esc_html__( 'Text for Contact Box', 'cmb2' ),
		'desc'    => esc_html__( 'add some copy for the contact box below the directors listing', 'cmb2' ),
		'id'      => 'contact_copy',
		'type'    => 'wysiwyg',
	) );
	$cmb->add_field( array(
		'name'    => esc_html__( 'Mailing Address', 'cmb2' ),
		'desc'    => esc_html__( 'add a mailing address for people to reach you', 'cmb2' ),
		'id'      => 'contact_mailing',
		'type'    => 'wysiwyg',
	) );
	$cmb->add_field( array(
		'name'    => esc_html__( 'Phone Number', 'cmb2' ),
		'desc'    => esc_html__( 'add a phone number', 'cmb2' ),
		'id'      => 'contact_phone',
		'type'    => 'text',
	) );
	$cmb->add_field( array(
		'name'    => esc_html__( 'Fax Number', 'cmb2' ),
		'desc'    => esc_html__( 'add a fax number', 'cmb2' ),
		'id'      => 'contact_fax',
		'type'    => 'text',
	) );
}
`

This is the code in my custom post type plugin page. Not cant figure out why its not saving to the admin area.

**Comment by tw2113** [comment]
Not presently able to recreate. I copy/pasted your code above, changed the post type slug to make it a submenu item of, and hit save with some field values. When the page refreshed, it loaded everything fine.

One question I guess I have is have you checked your database to see if there's any spot that it is in fact saving, despite not apparently making it back to the UI?

**Comment by TheNewGuy16** [comment]
Thanks for the reply. Yeah I checked the database and it is saving....I
have just moved it out of the submenu and made it its own menu and now it
is saving in the admin area....but really would like to figure out how to
make it a submenu.  I do have some other issue but I am new to doing case
studies and all that. But I will try and get these issues to you soon.

**Comment by tw2113** [comment]
Sounds good.

**Comment by thesquaremedia** [comment]
I was having a similar issue so I used this to at least show the correct saved item

`'default' => isset( get_option( 'option_name' )['field_name'] ) ? get_option( 'option_name' )['field_name']:'',`

**Comment by absolutelynotjames** [comment]
Hi there,

Sorry for bumping such an old thread, but this appears to be affecting us. For reference, I've got a Wordpress multisite running 5.2 locally, with the latest version (2.7) of CMB2.

Adding and saving standalone admin pages work as expected, but the moment 'parent_slug' is set to anything (such as 'themes.php' to add an options page under the appearance menu), the form is no longer able to save/display values.

This seems to affect standalone multisites, as well as adding a network-wide options page (again, a standalone network options page works great, but I try to set the parent_slug to something like 'themes.php', the options appear in the right menu but the form values will not save).

I've got all my debugging flags on in my wp_config, I'm using the Query Monitor plugin for debugging, and I'm not seeing anything notable under errors, even in my js console. Is there anything I can do begin debugging why this may be occuring?

Thanks!
